### PR TITLE
Add ModelSDF serializer

### PR DESCRIPTION
**Pull request by caguero**
# 🎉 New feature

## Summary

Adds a serializer to the `ModelSDF` component.

## Checklist
- [X] Signed all commits for DCO
- [X] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/gazebo/components/Model.hh [hunk 1]
```diff
@@ -29,14 +32,54 @@ namespace gazebo
...
+  class SdfModelSerializer
+  {
+    /// \brief Serialization for `sdf::Model`.
+    /// \param[in] _out Output stream.
+    /// \param[in] _time Model to stream
+    /// \return The stream.
+    public: static std::ostream &Serialize(std::ostream &_out,
+                const sdf::Model &_model)
+    {
+      _out << _model.Element()->GetParent()->ToString("");
```
[the remark this answers is not in the document]
Comment (caguero) [reply include/ignition/gazebo/components/Model.hh]: That was actually wrong, it shouldn't be used. It should be fixed in 63323e1.

File: test/integration/components.cc [hunk 2]
```diff
@@ -1099,6 +1100,54 @@ TEST_F(ComponentsTest, Model)
...
 
+/////////////////////////////////////////////////
+TEST_F(ComponentsTest, ModelSdf)
+{
+  std::ostringstream stream;
+  std::string version = SDF_VERSION;
+  stream
+    << "<?xml version=\"1.0\" ?>"
+    << "<sdf version='" << version << "'>"
+    << "<model name='my_model'>"
```
[the remark this answers is not in the document]
Comment (caguero) [reply test/integration/components.cc]: Done in 63323e1.

**Review by chapulina — APPROVED** [review]
LGTM :+1:

File: include/ignition/gazebo/components/Model.hh [hunk 3]
```diff
@@ -29,14 +32,64 @@ namespace gazebo
...
+    {
+      sdf::ElementPtr modelElem = _model.Element();
+      if (!modelElem)
+      {
+        ignerr << "Unable to serialize sdf::Model" << std::endl;
+        return _out;
+      }
+
+      _out << "<?xml version=\"1.0\" ?>"
+           << "<sdf version='" << modelElem->OriginalVersion() << "'>"
```
Comment (azeey) [inline include/ignition/gazebo/components/Model.hh:54]: The `OriginalVersion()` is the version of the file before it is (potentially) upconverted by the parser to the latest version. However, `ToString()` outputs the upconverted version, so I would say we should use `SDF_PROTOCOL_VERSION ` here instead.

File: include/ignition/gazebo/components/Model.hh [hunk 4]
```diff
@@ -29,14 +32,64 @@ namespace gazebo
...
+      std::string sdf(std::istreambuf_iterator<char>(_in), {});
+
+      sdf::Errors errors = root.LoadSdfString(sdf);
+      if (!errors.empty())
+      {
+        ignerr << "Unable to unserialize sdf::Model" << std::endl;
+        return _in;
+      }
+
+      _model.Load(root.Element()->GetElement("model"));
```
Comment (azeey) [inline include/ignition/gazebo/components/Model.hh:77]: I think doing `_model = *_root.Model()` should work and would avoid loading the model twice.

**Comment by caguero** [comment]
I'm considering closing this PR in favor of pull request #881. That way all the changes needed for "Scaling" will be contained in the `scale` branch.

**Comment by chapulina** [comment]
> I'm considering closing this PR in favor of pull request #881. That way all the changes needed for "Scaling" will be contained in the scale branch.

The changes here are pretty self-contained and independent of scaling, so I think it's ok to get it into `main`, then rebase the `scale` branch on top of `main`. But if you want to move it there I think it's ok too.

File: include/ignition/gazebo/components/Model.hh (hunk 3 above)
Comment (caguero) [reply include/ignition/gazebo/components/Model.hh]: Thanks, replaced in de79f9a

File: include/ignition/gazebo/components/Model.hh (hunk 4 above)
Comment (caguero) [reply include/ignition/gazebo/components/Model.hh]: Changed in de79f9a.

File: include/ignition/gazebo/components/Model.hh [hunk 5]
```diff
@@ -29,14 +32,64 @@ namespace gazebo
...
+    /// \param[out] _model Model to populate
+    /// \return The stream.
+    public: static std::istream &Deserialize(std::istream &_in,
+                sdf::Model &_model)
+    {
+      sdf::Root root;
+      std::string sdf(std::istreambuf_iterator<char>(_in), {});
+
+      sdf::Errors errors = root.LoadSdfString(sdf);
+      if (!errors.empty())
```
Comment (jennuine) [inline include/ignition/gazebo/components/Model.hh:71]: Should this check `!root.Element()->GetElement("model")` instead? https://github.com/ignitionrobotics/ign-gazebo/pull/881#discussion_r666595711

Comment (caguero) [reply include/ignition/gazebo/components/Model.hh]: Changed! c428d20

**Review by azeey — APPROVED** [review]

File: include/ignition/gazebo/components/Model.hh (hunk 5 above)
Comment (azeey) [reply include/ignition/gazebo/components/Model.hh]: Could also do `!root.Model()`

**Pull request merged**